### PR TITLE
Emerald: query token name etc

### PR DESCRIPTION
**Pull request by pro-wh**
we have a runtime client. let's run evm calls to query the ERC-20 `name` etc

### CODE REVIEW

**Comment by pro-wh** [comment]
got this convenient piece of josn from compiling openzeppelin erc20 in remix

```json
		"methodIdentifiers": {
			"allowance(address,address)": "dd62ed3e",
			"approve(address,uint256)": "095ea7b3",
			"balanceOf(address)": "70a08231",
			"burn(uint256)": "42966c68",
			"burnFrom(address,uint256)": "79cc6790",
			"decimals()": "313ce567",
			"decreaseAllowance(address,uint256)": "a457c2d7",
			"increaseAllowance(address,uint256)": "39509351",
			"mint(address,uint256)": "40c10f19",
			"name()": "06fdde03",
			"owner()": "8da5cb5b",
			"renounceOwnership()": "715018a6",
			"symbol()": "95d89b41",
			"totalSupply()": "18160ddd",
			"transfer(address,uint256)": "a9059cbb",
			"transferFrom(address,address,uint256)": "23b872dd",
			"transferOwnership(address)": "f2fde38b"
		}
```

File: analyzer/emerald/emerald.go [hunk 1]
```diff
@@ -270,23 +270,29 @@ func (m *Main) prepareBlockData(ctx context.Context, round uint64, batch *storag
...
 
-func (m *Main) queueBlockAndTransactionInserts(batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
+func (m *Main) queueBlockAndTransactionInserts(ctx context.Context, batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
 	blockData, err := extractRound(data.BlockHeader, data.TransactionsWithResults, m.logger)
 	if err != nil {
 		return fmt.Errorf("extract round %d: %w", data.Round, err)
 	}
 
+	// todo: move out of queue-* section
+	blockTokenData, err := downloadRoundTokens(ctx, m.logger, m.cfg.Source.(*oasis.RuntimeClient), data.Round, blockData.DirtyTokens)
```
Comment (pro-wh) [inline analyzer/emerald/emerald.go:291]: we had some ideas about doing this in a separate analyzer. the emerald scanner would record any tokens it encounters, then the other analyzer would read that list and run these online calls, then store the results similar to off-chain data

File: analyzer/emerald/incoming.go [hunk 2]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
 
-func emitRoundBatch(batch *storage.QueryBatch, qf *analyzer.QueryFactory, round uint64, blockData *BlockData) {
+var ABIERC20 abi.ABI
+
+func init() {
+	var err error
+	ABIERC20, err = abi.JSON(strings.NewReader(`
+  [
+    {
+      "constant": true,
```
Comment (pro-wh) [inline analyzer/emerald/incoming.go:442]: might be nice to have this compacted to be one line per abi entry. a middle ground between this menace to scrolling and an impenetrable single uglified line

File: analyzer/emerald/incoming_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/oasisprotocol/oasis-indexer/cmd/common"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis"
+)
+
+func TestTokenDataERC20(t *testing.T) {
+	// todo: can't keep this
+	ctx := context.Background()
+	cf, err := oasis.NewClientFactory(ctx, config.DefaultNetworks.All["mainnet"])
```
Comment (pro-wh) [inline analyzer/emerald/incoming_test.go:19]: let's delete this test for now.
after the mvp, let's come up with a way to deploy a token in our test fixture and reference that

File: analyzer/emerald/emerald.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/emerald/emerald.go:291]: oh also we gotta unhack that downcast :skull:

**Review by mitjat — COMMENTED** [review]
Thanks Warren. Nice job on figuring out how to talk to EVM.

I have a good number of comments, but they're largely about cleanup -- renames, tiny refactors, clarification requests, that sort of thing. Overall, this looks good!

File: analyzer/emerald/incoming_test.go (hunk 3 above)
Comment (pro-wh) [reply analyzer/emerald/incoming_test.go:19]: right, not offline. external dependencies

I've only used this to develop the reader, so that I could invoke the new code easily without actually scanning blocks

ok let's keep this. devs feel free to remove it if mainnet gets annihilated

File: analyzer/emerald/contracts/ERC20.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: CC0-1.0
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/contracts/ERC20.sol:1]: this is used to compile the artifacts/ERC20.json. use solc to compile.

contracts/ is source code, artifacts/ is compiled output. this seems to be convention in smart contract projects. let's stick to that if possible.

analyzer/ is fine. gonna try evm-interfaces

File: analyzer/emerald/incoming.go (hunk 2 above)
Comment (pro-wh) [reply analyzer/emerald/incoming.go:442]: that file doesn't get shipped with the compiled indexer program, and I don't want us to use "our docker build will still work" as an excuse to complicate the indexer program.

let's look for ways to embed the file at build time

File: analyzer/emerald/incoming.go [hunk 5]
```diff
@@ -55,14 +60,38 @@ type TokenChangeKey struct {
...
-	AddressPreimages map[string]*AddressPreimageData
-	TokenChanges     map[TokenChangeKey]*big.Int
+	Hash                string
+	NumTransactions     int
+	GasUsed             uint64
+	Size                int
+	TransactionData     []*BlockTransactionData
+	AddressPreimages    map[string]*AddressPreimageData
+	TokenBalanceChanges map[TokenChangeKey]*big.Int
+	// key is addr
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:82]: oasis bech32 encoded address https://github.com/oasisprotocol/oasis-indexer/blob/fcb0d1748653c8b435058c48a455597bc0c11a5d/analyzer/uncategorized/addresses.go#L14

gonna add that to the comment for now

pls add the type in a separate change uniformly after this

File: analyzer/emerald/incoming.go [hunk 6]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	ctx context.Context,
+	client *oasis.RuntimeClient,
+	round uint64,
+	contractEthAddr []byte,
+	contractABI *abi.ABI,
+	name string,
+	outI interface{},
+	in ...interface{},
+) error {
+	// https://github.com/oasisprotocol/emerald-web3-gateway/blob/v3.0.0/rpc/eth/api.go#L408
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:674]: oops I meant to expand that link to include the other choices too including gas price and caller and value

File: analyzer/emerald/incoming.go [hunk 7]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	inPacked, err := contractABI.Pack(name, in...)
+	if err != nil {
+		panic(err)
+	}
+	outPacked, err := client.EvmSimulateCall(ctx, round, gasPrice, gasLimit, caller, contractEthAddr, value, inPacked)
+	if err != nil {
+		return fmt.Errorf("runtime client evm simulate call: %w", err)
+	}
+	if err = contractABI.UnpackIntoInterface(outI, name, outPacked); err != nil {
+		return fmt.Errorf("unpack out: %w", err)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:689]: gonna have to write a longer response to this :skull:

File: analyzer/emerald/incoming.go [hunk 8]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+// nolint: unparam
+func evmCallWithABI(
+	ctx context.Context,
+	client *oasis.RuntimeClient,
+	round uint64,
+	contractEthAddr []byte,
+	contractABI *abi.ABI,
+	name string,
+	outI interface{},
+	in ...interface{},
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:672]: this is modeled after go-ethereum's Call https://github.com/ethereum/go-ethereum/blob/v1.10.19/accounts/abi/bind/base.go#L149

hm come to think of it, outI should come before name

`name` used here instead of `method` because that was what the abi `Pack` parameter called it \: and we don't use 'params' and 'results' either \: \:

File: analyzer/emerald/incoming.go [hunk 9]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	gasPrice := []byte{1}
+	gasLimit := uint64(30_000_000)
+	caller := ethCommon.Address{1}.Bytes()
+	value := []byte{0}
+
+	inPacked, err := contractABI.Pack(name, in...)
+	if err != nil {
+		panic(err)
+	}
+	outPacked, err := client.EvmSimulateCall(ctx, round, gasPrice, gasLimit, caller, contractEthAddr, value, inPacked)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:684]: :crossed_fingers: hoping that can give is some sanity when we have this code separated from the block scanning

File: analyzer/emerald/incoming.go [hunk 10]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	contractEthAddr []byte,
+	contractABI *abi.ABI,
+	name string,
+	outI interface{},
+	in ...interface{},
+) error {
+	// https://github.com/oasisprotocol/emerald-web3-gateway/blob/v3.0.0/rpc/eth/api.go#L408
+	gasPrice := []byte{1}
+	gasLimit := uint64(30_000_000)
+	caller := ethCommon.Address{1}.Bytes()
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:677]: using the same caller that the web3 gateway uses so that we don't have to decide for ourselves https://github.com/oasisprotocol/emerald-web3-gateway/blob/v3.0.0/rpc/eth/api.go#LL405

maybe we don't want it to act as unauthenticated? dunno

File: analyzer/emerald/incoming.go [hunk 11]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	value := []byte{0}
+
+	inPacked, err := contractABI.Pack(name, in...)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:682]: oh yeah it used to be statically known as there had been no inputs, but now the signature does allow inputs. let's return error :+1:

File: analyzer/emerald/incoming.go [hunk 12]
```diff
@@ -55,14 +60,38 @@ type TokenChangeKey struct {
...
+
+type TokenData struct {
+	Name        string
+	Symbol      string
+	Decimals    uint8
+	TotalSupply *big.Int
+}
+
+type BlockTokenData struct {
+	TokenData map[string]*TokenData
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:94]: ya data is too generic. let's see what the eip calls it

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:94]: oh they don't have a word for it. mostly stuff "to improve usability" other than the total supply.

'metadata' and 'description' would be a good fit, if it weren't for the total supply

File: analyzer/emerald/emerald.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/emerald/emerald.go:291]: k

File: analyzer/emerald/incoming.go [hunk 13]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "decimals", &tokenData.Decimals); err != nil {
+		logError("decimals", err)
+	}
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "totalSupply", &tokenData.TotalSupply); err != nil {
+		logError("totalSupply", err)
+	}
+	return &tokenData, nil
+}
+
+func downloadRoundTokens(ctx context.Context, logger *log.Logger, client *oasis.RuntimeClient, round uint64, tokens map[string]*DirtyToken) (*BlockTokenData, error) {
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:720]: ya more info:
- type column will be good. and it sounds like we'll be able to use EIP-165 for most of them
- NonToken -> Unrecognized probably, based on your description of it
- DirtyTokens and DirtyToken.Type: yeah good on possibleType being a list. or maybe some special enum with values like ObservedTransferAddressAddressUint256 :skull: but that's more complicated to program

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:720]: other thing: once we get this into a separate analyzer, it'll be able to check the db to see what it already has, e.g. not get name+symbol+decimals on tokens where we already have that data. tokens where that stuff changes dynamically, that's their problem :laughing:

File: analyzer/emerald/incoming.go [hunk 14]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+	var blockTokenData BlockTokenData
+	blockTokenData.TokenData = map[string]*TokenData{}
+	// todo: slow and serialized, lots of round trips
+	for tokenAddr, dirtyToken := range tokens {
+		switch dirtyToken.Type {
+		case TokenTypeERC20:
+			tokenData, err := downloadTokenERC20(ctx, logger, client, round, dirtyToken.EthAddr)
+			if err != nil {
+				return nil, fmt.Errorf("download token ERC20 %s %s (eth): %w", tokenAddr, hex.EncodeToString(dirtyToken.EthAddr), err)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:729]: "we found that it doesn't really implement this token standard" (A) is one outcome, and we'd go on to try other candidates or mark the token as unrecognized. however, and this may be a minor point, I'm not decided on using a Go error return to indicate that case.

other outcomes that'll be important:

"this was an ERC20, and we have its data" (B)

"this was an ERC20, but some optional methods weren't supported" (C)

"a network error prevented us from accessing this contract" (D)

D will be a Go error. A might or might not. C is close enough to B that we can likely have a struct with nullable fields.

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:729]: D we'll want to break out of this loop, unless maybe we want to retain results and save partial data to the db. we usually do this at larger granularity though e.g. per block in the scanner.

to handle A as an error, we might be able to play around with error.Is et al. or it might be simpler to have it be a kind of success

File: analyzer/emerald/incoming.go [hunk 15]
```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
...
+			}
+			blockTokenData.TokenData[tokenAddr] = tokenData
+		default:
+			return nil, fmt.Errorf("unrecognized token %s type %v", tokenAddr, dirtyToken.Type)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:733]: ought to panic

File: analyzer/queries.go [hunk 16]
```diff
@@ -429,14 +429,26 @@ func (qf QueryFactory) AddressPreimageInsertQuery() string {
...
 	return fmt.Sprintf(`
 		INSERT INTO %[1]s.%[2]s_token_balances (token_address, account_address, balance)
 			VALUES ($1, $2, $3)
 		ON CONFLICT (token_address, account_address) DO
 			UPDATE SET balance = %[1]s.%[2]s_token_balances.balance + $3`, qf.chainID, qf.runtime)
 }
 
+func (qf QueryFactory) RuntimeTokenUpdateQuery() string {
+	return fmt.Sprintf(`
+		INSERT INTO %s.%s_tokens (token_address, token_name, symbol, decimals, total_supply)
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/queries.go:442]: that we should have combined tables that included data from multiple paratimes (right?)--I'm in favor

File: analyzer/emerald/incoming.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:689]: my opinion when it comes to creating errors, wrapping errors, and handling errors

when wrapping an error in another error, only provide context. we _don't_ want error messages like "failed to do A: error in B: failed to do C: error in D." it's excessive to say something failed multiple times. instead, prefer to build error messages like "scan failed: in A: in B: in C: in D condition"

- the innermost error ought to have a condition stated as fact, with no interpretation about its supposed impact on the program
  - good: `fmt.Errorf("cache does not contain key %s", key)`
  - bad, 'crashing' inappropriate: `fmt.Errorf("crashing due to missing key %s", key)`
- when wrapping errors, provide context. identify where the error came from. again, don't include interpretation about its supposed impact on the program
  - good, in a function that looks up sender and receiver: `fmt.Errorf("looking up sender: %w", err)`
  - bad, missing context about which lookup failed: `fmt.Errorf("cache lookup: %w", err)`
  - good, in a function that loops through a transaction list: `fmt.Errorf("transaction %d getting affected accounts: %w", i, err)`
  - bad, can find source code line, but not iteration in loop: `fmt.Errorf("getting affected accounts: %w", err)`
  - bad, 'failed' inappropriate: `fmt.Errorf("failed to look up cached sender: %w", err)`
  - bad, throwing away error info: `fmt.Errorf("failed to look up cached sender", txHash)`
  - fine, in a straight-line function that only calls another function, so no additional context is needed: `err`
- when handling the error i.e. on the outermost layer, describe the impact to the program
  - good: `logger.Error("unable to process transaction, skipping", "err", err)`
  - bad, no description of impact: `logger.Error("processing transaction", "err", err)`
  - bad, throwing away error info `logger.Error("unable to process transaction, skipping")`
  - good, as panic already says that it is panicking: `panic(fmt.Sprintf("parsing hardcoded special address %s: %v", SpecialAddress, err))`

if A calls B, and B fails, who should say it was B? imo, A

- (i) "parsing To address: checksum mismatch" vs (ii) "To: parsing address checksum mismatch" I like (i)
- our codebase tends to have fewer callsites for each function and more places in each function that returns with error 

how detailed should the context be when wrapping:

- at a minimum, it should uniquely identify a location in the function, otherwise can be unclear what's failing
- when the control flow is complex, add information about that, e.g. iteration in a loop
- as mentioned by @mitjat, it can be helpful if it is unique in the whole project, so people can arrive by searching
- try to keep the context about what's inside the function, so that you don't need to adjust much when code paths leading to the function change

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:689]: so as it applies to this line:

- image
 by some luck, this really is unique in the codebase
- "unpack" is enough to identify UnpackIntoInterface as the source of the error
- incidentally there's only one call to UnpackIntoInterface, but specifying `out` might be helpful if later refactoring moves this code to somewhere with other similar calls
- identifying that it's "evm contract call" we should do in the caller of `evmCallWithABI`. currently that's only "ERC20 call failed." I don't mind adding that it's an EVM contract call there
- "failed to" I'm avoiding because the code that handles the error should already indicate that something failed
- the string here is short, so there's the fear that it won't be unique. but I don't want us to choose other conventions to make it longer. if we follow different conventions, it's about the same amount of information and would have the same likelihood of repeating in the codebase:
  - include "failed to": "unpack out" -> "failed to unpack out"
  - callee identifies itself instead of caller identifies callee: "failed to unpack out" -> "evm call failed, out"
- identifying the callee in both the caller an callee I think would give us a better chance at uniqueness, but it's repetitive in the resulting error message: "download round tokens downloading erc20: download erc20 making evm call for name: evm call for name unpacking out" :shrug:

File: analyzer/emerald/incoming.go (hunk 2 above)
Comment (pro-wh) [reply analyzer/emerald/incoming.go:442]: o crud I forgot that the artifact json is a larger object with a .abi property

Comment (pro-wh) [reply analyzer/emerald/incoming.go]: ok there's an easy enough way to say that we want just the .abi part, although we end up embedding the whole artifact json.

Comment (pro-wh) [reply analyzer/emerald/incoming.go]: changed to load from ERC20.json. at compile time

File: analyzer/emerald/incoming_test.go (hunk 3 above)
Comment (pro-wh) [reply analyzer/emerald/incoming_test.go:19]: changed commit message. filed #250

File: analyzer/emerald/contracts/ERC20.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/contracts/ERC20.sol:1]: analyzer/evm-interfaces/contracts/aritfacts/ERC20.json

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/contracts/ERC20.sol:1]: hm Go doesn't like packages names that way. analyzer/evmabi

File: analyzer/emerald/incoming.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: comment updated

File: analyzer/emerald/incoming.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: link fixed to include the other fields

File: analyzer/emerald/incoming.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: lengthened this and a few others

File: analyzer/emerald/incoming.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: returned

File: analyzer/emerald/incoming.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: changed to panic. can revisit after #225

File: analyzer/emerald/incoming.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: good discussion. I've linked here from #225. also added a comment to DirtyToken about future type change. closing to unblock.

File: analyzer/emerald/emerald.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/emerald/emerald.go]: hm moved to modules/evm.go. let's see how the team feels. it's a lot of exported symbols compared to most modules currently

File: analyzer/queries.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (gw0) [reply analyzer/queries.go:442]: On one hand, combining by SDK module makes sense. Adding new ParaTimes would just be a matter of right configuration.

On the other, grouping them by "analyzer instances" (I suppose this is the current situation) also makes sense. With this approach by looking at a table you know which "analyzer instance" generated it. If an "analyzer instance" starts having issues, you also know exactly which table is completely behind the latest round. In case of data corruption or the need to rollback, it is also much easier to rollback just one table or drop the table and re-run the "analyzer instance" again, instead of rolling back or dropping just a part of the table (huh!?).

File: analyzer/emerald/incoming.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: more opinions from outside: the abi json has fields 'inputs', 'name', 'outputs'

oasis-core uses 'method', 'args', 'rsp'

gonna try 'result', 'method', 'params' from go-ethereum's Call

File: analyzer/queries.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/queries.go:442]: would doing this interfere with grouping by "analyzer instances"? how do we do our databases for multiple analyzer instances currently?

[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/queries.go:442]: > On one hand, combining by SDK module makes sense. Adding new ParaTimes would just be a matter of right configuration.
This is a strong motivating factor IMO, especially given the similarities between Emerald and Sapphire (and the desire to launch Sapphire pretty fast).

 For your other points, @gw0: True, but mostly negated by the fact that an "analyzer instance" (these are per-paratime, if you count consensus as a paratime; you could have multiple instances per paratime, but then the current db design doesn't separate their data either) will always write the paratime name into a column of all these paratime-shared tables. So "dropping" becomes a little more complicated ("DELETE FROM ... WHERE paratime=...") but still doable in a pinch. And for local testing or total from-scratch redeploys (= most common use cases, I expect), it's still OK to go nuclear and wipe all the analyzers' state.

File: analyzer/emerald/incoming.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/incoming.go]: I think I wasn't very clear in my original comment; I was (also?) advocating for this not being a Go error in case A. Good point that D exists and is a very different failure. Yeah I agree with the `// todo` that nondeterministic errors (meaning type D) should be propagated out of `evmDownloadTokenERC20()`, and other errors should just impact setting the token type. IIRC you said in a different comment thread that you want to do that in a separate PR? Either way.

File: analyzer/queries.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (gw0) [reply analyzer/queries.go:442]: Oh, if the current db design does not separate paratime data, then I guess it is time to change... :)

After sleeping over it, I believe combining by SDK module and storing the paratime name into a column is actually much better. Potential bugs that need dropping and re-indexing are probably contained within an SDK module, and re-running just one such analyzer for a paratime should be much faster than all of them. Thanks for clarifying!

File: analyzer/emerald/contracts/ERC20.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/contracts/ERC20.sol]: > contracts/ is source code, artifacts/ is compiled output. this seems to be convention in smart contract projects. let's stick to that if possible.

Gotcha, thanks. Can we add that into code somewhere? A separate `README.md` in `evmabi/`, or a top-level comment in `evmabi.go`, perhaps.

File: analyzer/emerald/incoming.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/incoming.go]: I think "unpack out" made me comment because it's so ungrammatical / hard to parse without context. It wasn't clear that "out" is a variable name and not a weird preposition. I guess I touched on something :). Anyway, I like what you landed on. Thank you :) 

I agree with your guidelines almost entirely. The only slight difference in opinion is that in an ideal world, a function is self-contained and should not / cannot know or care whether it's the the outermost layer of error message compositing. If you subscribe to that idea, it encourages more verbosity at the cost of some repetition in the final composed message (which I personally also don't mind much).

The last comment made me laugh. I emotionally relate to the process :)
"Somebody is Wrong on the Internet™" -> over-engage -> "ya alright let's do it your way, let's wrap this up, it doesn't matter much"

File: analyzer/emerald/incoming.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/incoming.go]: Can't say I love it but I can't think of a clear winner either; fair point with total supply being the odd one out. Thanks for exploring.

File: analyzer/emerald/incoming.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/incoming.go]: yep good additional points :+1: 

sg for NonToken -> Unrecognized, that's probably a better name

Don't you want to implement `Unrecognized` now? Oh well I guess it goes hand in hand with the next comment thread, handling network errors when fetching token data, which might happen in a separate PR.

Unresolving for visibility only, feel free to resolve if you're doing this in a separate PR.

File: analyzer/emerald/incoming.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (mitjat) [reply analyzer/emerald/incoming.go]: We'll hopefully never hit this, so it doesn't matter.

But on the off-chance we do hit it, the difference between `Error` and `panic` is "slightly corrupt DB that you can fix over the next few days" vs "pagerduty in the night".

For right now, panic is good enough for sure. Might be nice to leave a TODO to remind yourself when working on #225.

**Review by mitjat — APPROVED** [review]
Thank you for all the additional work on this. There are some new/remaining comment threads from my side but they should be small and noncontroversial. Once resolved, LGTM!

File: analyzer/emerald/contracts/ERC20.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/contracts/ERC20.sol]: leme add a README, yeah

File: analyzer/queries.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (gw0) [reply analyzer/queries.go:442]: In the `operators-health` project table names are a configurable parameter for each "runner" (the equivalent of analyzer). Not sure if it is the best way, but it is definitely the most flexible way.

[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/queries.go:428]: :+1: separate PR is fine

File: analyzer/modules/evm.go [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer/evmabi"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+)
+
+type EVMTokenType int
+
+const (
+	EVMTokenTypeERC20 EVMTokenType = iota
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go:20]: k

File: analyzer/emerald/incoming.go [hunk 18]
```diff
@@ -56,13 +57,15 @@ type TokenChangeKey struct {
...
-	TokenChanges     map[TokenChangeKey]*big.Int
+	Hash                string
+	NumTransactions     int
+	GasUsed             uint64
+	Size                int
+	TransactionData     []*BlockTransactionData
+	AddressPreimages    map[string]*AddressPreimageData
+	TokenBalanceChanges map[TokenChangeKey]*big.Int
+	// key is oasis bech32 address
+	DirtyTokens map[string]*modules.EVMDirtyToken
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:68]: ok on the rename

clarification on the addr repetition: map key is oasis bech32 addr, struct field is eth addr binary. this so we don't have to trouble the db with preimage queries during the processing

File: analyzer/modules/evm.go [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+	Symbol      string
+	Decimals    uint8
+	TotalSupply *big.Int
+}
+
+type EVMBlockTokenData struct {
+	TokenData map[string]*EVMTokenData
+}
+
+// nolint: unparam
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go:40]: currently all callers pass no inputs, so the linter would say to remove the param altogether

File: analyzer/modules/evm.go [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+func evmCallWithABI(
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go:41]: thanks, I'll add that

File: storage/migrations/02_oasis_3_emerald.up.sql [hunk 21]
```diff
@@ -88,26 +88,35 @@ CREATE INDEX ix_emerald_related_transactions_address_height_index ON oasis_3.eme
...
+  account_address oasis_addr NOT NULL,
   balance NUMERIC(1000,0) NOT NULL,
   PRIMARY KEY (token_address, account_address)
 );
 
 CREATE INDEX ix_emerald_token_address ON oasis_3.emerald_token_balances (token_address) WHERE balance != 0;
 
+CREATE TABLE oasis_3.emerald_tokens
+(
+  token_address TEXT NOT NULL PRIMARY KEY,
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply storage/migrations/02_oasis_3_emerald.up.sql:113]: :+1:

File: analyzer/emerald/incoming.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go:68]: or PossibleTypes, I'm thinking. as it's in a struct already pertaining to a token

File: analyzer/emerald/contracts/ERC20.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/contracts/ERC20.sol]: added

File: analyzer/emerald/incoming.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: closing for merge. thanks for the comment

File: analyzer/emerald/incoming.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: we'll be going on to use the same settings as the web3 gateway. thanks for asking

File: analyzer/emerald/incoming.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: I'll be taking on the Unrecognized along with the deterministic error handling in a new PR :skull: 
resolving for merge

File: analyzer/emerald/incoming.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/incoming.go]: will be carrying on in a separate PR. thanks for the ideas. resolving for merge

File: analyzer/queries.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/queries.go:428]: filed 262, thanks for bringing this up. closing for merge

File: analyzer/modules/evm.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go]: edited EvmTokenTypeERC20 to 20.

haven't added 721 yet, let's do that within the realm of #225

File: analyzer/emerald/emerald.go [hunk 22]
```diff
@@ -272,23 +272,29 @@ func (m *Main) prepareBlockData(ctx context.Context, round uint64, batch *storag
...
 }
 
-func (m *Main) queueBlockAndTransactionInserts(batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
+func (m *Main) queueBlockAndTransactionInserts(ctx context.Context, batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
 	blockData, err := extractRound(data.BlockHeader, data.TransactionsWithResults, m.logger)
 	if err != nil {
 		return fmt.Errorf("extract round %d: %w", data.Round, err)
 	}
 
+	// todo: move out of queue-* section
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/emerald/emerald.go:292]: filed #263, added reference in comment

File: analyzer/modules/evm.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go:45]: thanks for asking. resolving for merge

File: analyzer/modules/evm.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply analyzer/modules/evm.go:44]: added
